File: Good Travels.cpp
// https://dmoj.ca/problem/gooda
// python code has explanation and stuff
// note that case 6 is massive and overflows int (but we can't use long long for everything or else MLE)

#include <bits/stdc++.h>
#define ll long long
using namespace std;

vector<int> scc, low_link, node_id, fun;
vector<ll> cache, scc_fun;
vector<vector<int>> graph, compressed;
vector<bool> in_path;
stack<int> path;
int t = 0, comp = 1, N, M, start, _end;

void dfs(int cur) {
    node_id[cur] = low_link[cur] = t++;
    path.push(cur);
    in_path[cur] = true;

    for (int adj : graph[cur]) {
        if (node_id[adj] == -1) {
            dfs(adj);
            low_link[cur] = min(low_link[cur], low_link[adj]);
        } else if (in_path[adj]) {
            low_link[cur] = min(low_link[cur], node_id[adj]);
        }
    }

    if (node_id[cur] == low_link[cur]) {
        while (path.top() != cur) {
            in_path[path.top()] = false;
            scc_fun[comp] += fun[path.top()];
            scc[path.top()] = comp;
            path.pop();
        }
        in_path[cur] = false;
        scc_fun[comp] += fun[cur];
        scc[path.top()] = comp;
        path.pop();
        comp++;
    }
}

pair<vector<int>, vector<ll>> get_scc() {
    for (int i = 1; i <= N; i++) {
        if (node_id[i] == -1) {
            dfs(i);
        }
    }
    return make_pair(scc, scc_fun);
}

ll solve(ll cur) {
    if (cache[cur] != -1) {
        return cache[cur];
    }
    if (cur == scc[_end]) {
        return scc_fun[scc[_end]];
    }
    ll best = 0;
    for (int adj : compressed[cur]) {
        best = max(best, solve(adj) + (ll)scc_fun[cur]);
    }
    cache[cur] = best;
    return cache[cur];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N >> M >> start >> _end;
    fun.resize(N + 1, -1);
    for (int i = 1; i <= N; i++) {
        cin >> fun[i];
    }

    graph.resize(N + 1);
    for (int i = 0; i < M; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
    }

    scc.resize(N + 1, -1);
    scc_fun.resize(N + 1, 0);
    low_link.resize(N + 1, -1);
    node_id.resize(N + 1, -1);
    in_path.resize(N + 1, false);
    tie(scc, scc_fun) = get_scc();

    compressed.resize(comp);
    for (int i = 1; i <= N; i++) {
        for (int adj : graph[i]) {
            if (scc[i] != scc[adj]) {
                compressed[scc[i]].push_back(scc[adj]);
            }
        }
    }

    for (auto &adj : compressed) {
        unordered_set<int> s(adj.begin(), adj.end());
        adj.assign(s.begin(), s.end());
    }

    cache.resize(N + 1, -1);
    cout << solve(scc[start]) << "\n";

    return 0;
}

File: Good Travels.py
"""
6/7 MLE, check C++ code
https://dmoj.ca/problem/gooda

Observation:
- if we can reach a node in a strongly connected component (SCC), we can get the fun values of all other
  nodes in this SCC, which motivates us to turn SCCs into a single node
- after compressing SCCs into a single node (with tarjan's algorithm) , we are left with a DAG
  - this allows us to use graph DP
"""

import sys

input = sys.stdin.readline
sys.setrecursionlimit(200000)


def get_scc(graph):
    n = len(graph)

    scc = [-1] * n  # the component that node i belongs to
    scc_fun = [0] * n  # sum of fun values in this SCC
    comp = 1  # current component
    low_link = [-1] * n  # smallest time reachable of a node reachable from the DFS subtree of node i
    node_id = [-1] * n  # time at which node i was discovered
    time = 0
    path = []  # nodes currently in the dfs traversal
    in_path = [False] * n

    def dfs(cur):
        nonlocal time, comp
        node_id[cur] = low_link[cur] = time
        path.append(cur)
        in_path[cur] = True
        time += 1

        for adj in graph[cur]:
            if node_id[adj] == -1:  # not yet visited
                dfs(adj)
                low_link[cur] = min(low_link[cur], low_link[adj])

            elif in_path[adj]:  # back edge, since we arrived at a node on the current path
                low_link[cur] = min(low_link[cur], node_id[adj])

            else:
                # 1-way path to another component, NOT back edge
                continue

        if node_id[cur] == low_link[cur]:  # root in dfs tree, assign SCC values to everything in this SCC
            while path[-1] != cur:
                in_path[path[-1]] = False
                scc_fun[comp] += fun[path[-1]]
                scc[path.pop()] = comp
            in_path[cur] = False
            scc_fun[comp] += fun[cur]
            scc[path.pop()] = comp
            comp += 1

    for i in range(1, n):
        if node_id[i] == -1:
            dfs(i)
    return scc, scc_fun


N, M, start, end = map(int, input().split())
fun = [-1] + [int(input()) for _ in range(N)]

graph = [[] for _ in range(N + 1)]
for _ in range(M):
    a, b = map(int, input().split())
    graph[a].append(b)

# compress SCCs into a single node with the sum of their fun values
scc, scc_fun = get_scc(graph)
compressed = [[] for _ in range(max(scc) + 1)]
for i in range(1, len(graph)):
    for adj in graph[i]:
        if scc[i] != scc[adj]:
            compressed[scc[i]].append(scc[adj])
compressed = list(map(lambda x: list(set(x)), compressed))  # remove duplicates

# the compressed graph is now a DAG, run simple graph DP on it to get maximum fun
cache = [-1] * len(compressed)


def solve(cur):
    if cache[cur] != -1:
        return cache[cur]
    if cur == scc[end]:  # reached end SCC
        return scc_fun[scc[end]]
    best = 0
    for adj in compressed[cur]:
        best = max(best, solve(adj) + scc_fun[cur])  # go to next SCC, add current SCC's fun value
    cache[cur] = best
    return cache[cur]


print(solve(scc[start]))
